File: src/boot/atags/atags.c
#include "atags.h"
#include "ros_kernel.h"
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>

#define ATAG_NONE      0
#define ATAG_CORE      0x54410001
#define ATAG_MEM       0x54410002
#define ATAG_VIDEOTEXT 0x54410003
#define ATAG_RAMDISK   0x54410004
#define ATAG_INITRD2   0x54410005
#define ATAG_SERIAL    0x54410006
#define ATAG_REVISION  0x54410007
#define ATAG_VIDEOLFB  0x54410008
#define ATAG_CMDLINE   0x54410009

#define count_of(a) (sizeof(a) / sizeof((a)[0]))
#define offset_ptr(p, s) ((const void*)(((const uint8_t*)(p)) + (s)))

/**
 * @struct ATAGHeader
 * @brief  Header descriptor for tags.
 */
typedef struct {
  uint32_t size; // Size of tag in 32-bit words
  uint32_t tag;  // Tag identifier
} ATAGHeader;

/**
 * @struct ATAGCore
 * @brief  Core kernel parameters.
 */
typedef struct {
  uint32_t flags;
  uint32_t page_size;
  uint32_t root_dev;
} ATAGCore;

/**
 * @struct ATAGMem
 * @brief  Memory region available to the kernel.
 */
typedef struct {
  uint32_t size; // Size of the region in bytes.
  uint32_t base; // Base address of the region.
} ATAGMem;

/**
 * @struct ATAG
 * @brief  Wrapper struct for a tag.
 */
typedef struct {
  ATAGHeader hdr;
  union {
    ATAGCore core;
    ATAGMem mem;
  } tag;
} ATAG;

static void handle_mem(ROSKernelInit *init, const ATAGMem *mem);

bool read_atags(ROSKernelInit *init, uintptr_t start) {
  const ATAG *p = (const ATAG*)start;

  if (p == NULL) {
    return false;
  }

  // A valid ATAG list must start with CORE.
  if (p->hdr.tag != ATAG_CORE) {
    return false;
  }

  while (p->hdr.tag != ATAG_NONE) {
    switch (p->hdr.tag) {
      case ATAG_MEM:
        handle_mem(init, &p->tag.mem);
        break;
      default:
        break;
    }

    p = offset_ptr(p, p->hdr.size * sizeof(uint32_t));
  }

  return true;
}

/**
 * @fn void handle_mem(ROSKernelInit *init, const ATAGMem *mem)
 * @brief Adds a memory region to the initialization struct.
 * @param[out] init The ROS kernel initialization struct.
 * @param[in] mem   The memory region ATAG.
 */
static void handle_mem(ROSKernelInit *init, const ATAGMem *mem) {
  static const size_t max_regions = count_of(init->mem_regions);
  size_t mem_rgn = 0;

  while (mem_rgn < max_regions && init->mem_regions[mem_rgn].size != 0) {
    ++mem_rgn;
  }

  if (mem_rgn == max_regions) {
    return;
  }

  init->mem_regions[mem_rgn].base = mem->base;
  init->mem_regions[mem_rgn].size = mem->size;
}

File: src/kernel/src/support/atags.rs
use super::kernel_init::ROSKernelInit;
use core::mem::ManuallyDrop;

const ATAG_NONE: u32 = 0;
const ATAG_CORE: u32 = 0x54410001;
const ATAG_MEM: u32 = 0x54410002;
const _ATAG_VIDEOTEXT: u32 = 0x54410003;
const _ATAG_RAMDISK: u32 = 0x54410004;
const _ATAG_INITRD2: u32 = 0x54410005;
const _ATAG_SERIAL: u32 = 0x54410006;
const _ATAG_REVISION: u32 = 0x54410007;
const _ATAG_VIDEOLFB: u32 = 0x54410008;
const _ATAG_CMDLINE: u32 = 0x54410009;

/// @struct ATAGHeader
/// @brief Header for all ATAG entries.
/// @var size Number of 32-bit integers in the ATAG.
/// @var tag  The tag identifier.
#[repr(C)]
struct ATAGHeader {
  size: u32,
  tag: u32,
}

/// @struct ATAGCore
/// @brief The CORE ATAG.
/// @var flags     ??? flags that indicate, you know, things ???
/// @var page_size ??? virtual memory page size ???
/// @var root_dev  ??? the root device something something ???
#[repr(C)]
struct ATAGCore {
  flags: u32,
  page_size: u32,
  root_dev: u32,
}

/// @struct ATAGMem
/// @brief The MEM ATAG.
/// @var size Size of the memory region in bytes.
/// @var base Base address of the region.
#[repr(C)]
struct ATAGMem {
  size: u32,
  base: u32,
}

/// @struct ATAGData
/// @brief Overlay of tag data.
#[repr(C)]
union ATAGData {
  core: ManuallyDrop<ATAGCore>,
  mem: ManuallyDrop<ATAGMem>,
}

/// @struct ATAG
/// @brief Full ATAG with header and overlaid data.
/// @var hdr The ATAG header.
/// @var tag Overlaid ATAG data.
#[repr(C)]
struct ATAG {
  hdr: ATAGHeader,
  tag: ATAGData,
}

/// @fn read_atags(init: &mut ROSKernelInit, blob: usize) -> bool
/// @brief   Read ATAGs provided by the bootloader.
/// @param[in] init The kernel initialization struct to fill out.
/// @param[in] blob Pointer to the ATAGs blob.
/// @returns True if able to read ATAGs, false if the blob is not an ATAG list.
pub fn read_atags(init: &mut ROSKernelInit, blob: usize) -> bool {
  unsafe {
    let mut ptr = blob as *const u32;
    let mut hdr = ptr as *const ATAGHeader;

    // The ATAG list must start with CORE.
    if (*hdr).tag != ATAG_CORE {
      return false;
    }

    loop {
      let atag = hdr as *const ATAG;

      match (*atag).hdr.tag {
        ATAG_NONE => break,
        ATAG_MEM => read_mem_atag(init, &(*atag).tag.mem),
        _ => {}
      }

      // Offset ptr. The size field is the number of 32-bit integers in the tag.
      ptr = ptr.offset((*atag).hdr.size as isize);
      hdr = ptr as *const ATAGHeader;
    }
  }

  true
}

/// @fn read_mem_atag(init: &mut ROSKernelInit, tag: &ATAGMem)
/// @brief Add a memory region to the kernel initialization struct.
/// @param[in] init The kernel initialization struct.
/// @param[in] tag  The memory region ATAG.
fn read_mem_atag(init: &mut ROSKernelInit, tag: &ATAGMem) {
  for mut rgn in init.memory_regions {
    if rgn.size == 0 {
      rgn.size = tag.size as usize;
      rgn.base = tag.base as usize;
      break;
    }
  }
}
